node/impl/full/f3.go

```diff
@@ -23,67 +22,74 @@ type F3API struct {
 	F3 *lf3.F3 `optional:"true"`
 }
 
-var ErrF3Disabled = errors.New("f3 is disabled")
-
-func (f3api *F3API) F3Participate(ctx context.Context, miner address.Address,
-	newLeaseExpiration time.Time, oldLeaseExpiration time.Time) (bool, error) {
-
-	if leaseDuration := time.Until(newLeaseExpiration); leaseDuration > 5*time.Minute {
-		return false, xerrors.Errorf("F3 participation lease too long: %v > 5 min", leaseDuration)
-	} else if leaseDuration < 0 {
-		return false, xerrors.Errorf("F3 participation lease is in the past: %d < 0", leaseDuration)
-	}
-
+func (f3api *F3API) F3GetOrRenewParticipationToken(ctx context.Context, miner address.Address, previous api.F3ParticipationToken) (api.F3ParticipationToken, error) {
 	if f3api.F3 == nil {
-		log.Infof("F3Participate called for %v, F3 is disabled", miner)
-		return false, ErrF3Disabled
+		log.Infof("F3GetParticipationToken called for %v, F3 is disabled", miner)
+		return api.F3ParticipationToken{}, api.ErrF3Disabled{}
 	}
 	minerID, err := address.IDFromAddress(miner)
 	if err != nil {
-		return false, xerrors.Errorf("miner address is not of ID type: %v: %w", miner, err)
+		return api.F3ParticipationToken{}, xerrors.Errorf("miner address is not of ID type: %v: %w", miner, err)
 	}
+	return f3api.F3.GetOrRenewParticipationToken(ctx, minerID, previous)
+}
+
+func (f3api *F3API) F3Participate(ctx context.Context, token api.F3ParticipationToken) (api.F3ParticipationLease, error) {
 
-	return f3api.F3.Participate(ctx, minerID, newLeaseExpiration, oldLeaseExpiration), nil
+	if f3api.F3 == nil {
+		log.Infof("F3Participate called, F3 is disabled")
+		return api.F3ParticipationLease{}, api.ErrF3Disabled{}
+	}
+	return f3api.F3.Participate(ctx, token)
 }
 
 func (f3api *F3API) F3GetCertificate(ctx context.Context, instance uint64) (*certs.FinalityCertificate, error) {
 	if f3api.F3 == nil {
-		return nil, ErrF3Disabled
+		return nil, api.ErrF3Disabled{}
 	}
 	return f3api.F3.GetCert(ctx, instance)
 }
 
 func (f3api *F3API) F3GetLatestCertificate(ctx context.Context) (*certs.FinalityCertificate, error) {
 	if f3api.F3 == nil {
-		return nil, ErrF3Disabled
+		return nil, api.ErrF3Disabled{}
 	}
 	return f3api.F3.GetLatestCert(ctx)
 }
 
-func (f3api *F3API) F3GetManifest(_ctx context.Context) (*manifest.Manifest, error) {
+func (f3api *F3API) F3GetManifest(context.Context) (*manifest.Manifest, error) {
 	if f3api.F3 == nil {
-		return nil, ErrF3Disabled
+		return nil, api.ErrF3Disabled{}
 	}
 	return f3api.F3.GetManifest(), nil
 }
 
-func (f3api *F3API) F3IsRunning(_ctx context.Context) (bool, error) {
+func (f3api *F3API) F3IsRunning(context.Context) (bool, error) {
 	if f3api.F3 == nil {
-		return false, ErrF3Disabled
+		return false, api.ErrF3Disabled{}
 	}
 	return f3api.F3.IsRunning(), nil
 }
 
 func (f3api *F3API) F3GetECPowerTable(ctx context.Context, tsk types.TipSetKey) (gpbft.PowerEntries, error) {
 	if f3api.F3 == nil {
-		return nil, ErrF3Disabled
+		return nil, api.ErrF3Disabled{}
```

Comment:
What if we:

1. Make this error type private.
2. Expose a public instance of it (`var ErrF3Disabled errF3Disabled`)?

I _think_ we'd still be able to use it with jsonrpc and it would make this code a little cleaner.

Reply:
That was my initial thought but was unsure if it would play nice with RPC errors. Going to give this a go.